README.rst
JSON-able data types
====================

This library provides and abstract base class ``JSONable`` which enables easy definition of trivially JSON-able python objects.

* **Installation:** ``pip install jsonable``

:Example:
    
    >>> from jsonable import JSONable
    >>>
    >>>
    >>> class Fruit(JSONable):
    ...     __slots__ = ('type', 'weight')
    ...
    ...     def initialize(self, type, weight):
    ...         self.type   = str(type)
    ...         self.weight = float(weight)
    ...
    >>> class Pie(JSONable):
    ...     __slots__ = ('fruit',)
    ...
    ...     def initialize(self, fruit):
    ...         self.fruit = [Fruit(f) for f in fruit]
    ...
    ...
    >>> pie = Pie([Fruit('apple', 10.3), Fruit('cherry', 2)])
    >>>
    >>> doc = pie.to_json()
    >>> doc
    {'fruit': [{'weight': 10.3, 'type': 'apple'}, {'weight': 2.0, 'type': 'cherry'}]}
    >>>
    >>> pie == Pie(doc)
    True


CHANGE_LOG.rst
v0.1.1
======
Fixes pickling issue using pickle protocol 3.

v0.1.0
======
Adds abstract JSONable class to handle abstract reconstruction.

v0.0.4
======
Added mangled _from_json() to preserve cross-class method visibility.

v0.0.3
======
Adds 'bool' to natively JSONable types.

v0.0.2
======
Fixed a minor bug in repr(JSONable)


jsonable/tests/test_instance.py
from nose.tools import eq_

from .. import instance


def test_simple_repr():
    
    eq_(
        instance.simple_repr("Classname", 'foo', 5, herp=[5]),
        "Classname('foo', 5, herp=[5])"
    )
    
    eq_(
        instance.simple_repr("Classname",
                             ordered_kwargs=[('foo', "foo"),
                                             ('five', 5),
                                             ('herp', [5]),
                                             ('derp', 20)]),
        "Classname(foo='foo', five=5, herp=[5], derp=20)"
    )

def test_slots_repr():
    
    class AbstractSlottedItem:
        __slots__ = ('foo', 'five')
        def __init__(self, foo, five):
            self.foo = foo
            self.five = five
        
    
    class SubSlottedItem(AbstractSlottedItem):
        __slots__ = ('herp', 'derp')
        def __init__(self, foo, five, herp, derp):
            super().__init__(foo, five)
            self.herp = herp
            self.derp = derp
        
    
    eq_(
        instance.slots_repr(SubSlottedItem("foo", 5, [5], 20)),
        "SubSlottedItem(foo='foo', five=5, herp=[5], derp=20)"
    )

def test_slots_items():
    
    class AbstractSlottedItem:
        __slots__ = ('foo', 'five')
        def __init__(self, foo, five):
            self.foo = foo
            self.five = five
        

    class SubSlottedItem(AbstractSlottedItem):
        __slots__ = ('herp', 'derp')
        def __init__(self, foo, five, herp, derp):
            super().__init__(foo, five)
            self.herp = herp
            self.derp = derp
        
    
    eq_(
        list(instance.slots_items(SubSlottedItem("foo", 5, [5], 20))),
        [("foo", "foo"), ("five", 5), ("herp", [5]), ("derp", 20)]
    )

def test_slots_keys():
    
    class AbstractSlottedItem:
        __slots__ = ('foo', 'five')
        def __init__(self, foo, five):
            self.foo = foo
            self.five = five
        

    class SubSlottedItem(AbstractSlottedItem):
        __slots__ = ('herp', 'derp')
        def __init__(self, foo, five, herp, derp):
            super().__init__(foo, five)
            self.herp = herp
            self.derp = derp
        
    
    eq_(
        list(instance.slots_keys(SubSlottedItem("foo", 5, [5], 20))),
        ["foo", "five", "herp", "derp"]
    )


jsonable/tests/test_jsonable.py
import pickle

from nose.tools import eq_

from ..jsonable import AbstractJSONable, JSONable


class ExampleJSONable(JSONable):
    __slots__ = ('foo', 'bar')
    def initialize(self, foo, bar):
        self.foo = foo
        self.bar = bar

def test_construction_and_variables():
    
    class Bar(JSONable):
        __slots__ = ('subherp', 'subderp')
        def initialize(self, subherp, subderp):
            self.subherp = subherp
            self.subderp = subderp
    
    class Foo(JSONable):
        __slots__ = ('herp', 'bars')
        def initialize(self, herp, bars):
            self.herp = herp
            self.bars = list(Bar(b) for b in bars)
        
    herp = 5
    bars = [Bar("string", 334.34), Bar(False, {"derp": 3.1})]
    foo = Foo(herp, bars)
    
    eq_(foo.herp, herp)
    eq_(foo.bars, bars)
    eq_(foo, Foo(foo))
    eq_(foo, Foo(foo.to_json()))
    

def test_pickle():
    bar = ExampleJSONable("wat", "herp")

    eq_(bar, pickle.loads(pickle.dumps(bar)))


def test_repr():
    
    class Bar(JSONable):
        __slots__ = ('subherp', 'subderp')
        def initialize(self, subherp, subderp):
            self.subherp = subherp
            self.subderp = subderp
        
    
    bar = Bar(1, "two")
    eq_(
        repr(bar),
        "Bar(subherp=1, subderp='two')"
    )

def test_abstract_construction_and_variables():
    
    class Bowl(JSONable):
        __slots__ = ('fruit',)
        def initialize(self, fruit):
            self.fruit = [Fruit(f) for f in fruit]
        
    
    
    class Fruit(AbstractJSONable):
        __slots__ = ('weight',)
        def initialize(self, weight):
            self.weight = float(weight) # lbs
            
    
    class Apple(Fruit):
        __slots__ = ('variety',)
        def initialize(self, weight, variety):
            super().initialize(weight)
            self.variety = str(variety)
        
    
    class Orange(Fruit):
        __slots__ = ('radius',)
        def initialize(self, weight, radius):
            super().initialize(weight)
            self.radius = float(radius) # in
            
        
    orange = Orange(10.1, 2.5)
    apple = Apple(9.2, "Honey Crisp")
    
    bowl = Bowl([apple, orange])
    print(bowl.to_json())
    
    
    fruit = Fruit(21.1)
    eq_(fruit, Fruit(fruit.to_json()))


jsonable/instance.py
import inspect
from itertools import chain


def simple_repr(class_name, *args, ordered_kwargs=None, **kwargs):
    class_name = str(class_name)
    ordered_kwargs = ordered_kwargs or []
    
    arguments = [repr(arg) for arg in args]
    arguments.extend("{0}={1}".format(k, repr(v)) for k, v in ordered_kwargs)
    arguments.extend("{0}={1}".format(k, repr(v)) for k, v in kwargs.items())
    
    return "{0}({1})".format(class_name, ", ".join(arguments))

def unique(l):
    seen = set()
    for i in l:
        if i not in seen:
            yield i
            seen.add(i)

def slots_repr(instance):
    
    return simple_repr(
        instance.__class__.__name__,
        ordered_kwargs = slots_items(instance)
    )

def slots_items(instance):
    for key in slots_keys(instance):
        yield key, getattr(instance, key)

def slots_values(instance):
    for key in slots_keys(instance):
        yield getattr(instance, key)

def slots_keys(instance):
    cls = instance.__class__
    
    return unique(chain(*(getattr(cls, '__slots__', [])
                     for cls in reversed(instance.__class__.__mro__))))


jsonable/jsonable.py
import copy

from . import instance
from .self_constructor import SelfConstructor

JSON_TYPES = {str, int, float, type(None), bool}


class JSONable(SelfConstructor):
    """
    Implements a simple class interface for trivially JSONable objects.
    
    :Example:
        >>> from jsonable import JSONable
        >>>
        >>>
        >>> class Fruit(JSONable):
        ...     __slots__ = ('type', 'weight')
        ...
        ...     def initialize(self, type, weight):
        ...         self.type   = str(type)
        ...         self.weight = float(weight)
        ...
        >>> class Pie(JSONable):
        ...     __slots__ = ('fruit',)
        ...
        ...     def initialize(self, fruit):
        ...         self.fruit = [Fruit(f) for f in fruit]
        ...
        ...
        >>> pie = Pie([Fruit('apple', 10.3), Fruit('cherry', 2)])
        >>>
        >>> doc = pie.to_json()
        >>>
        >>> pie == Pie(doc)
        True
    """
    def __new__(cls, *args, **kwargs):
        if len(args) == 1 and len(kwargs) == 0 and isinstance(args[0], dict):
            return cls.from_json(args[0])
        else:
            return super().__new__(cls, *args, **kwargs)
    
    def __eq__(self, other):
        if other == None: return False
        try:
            for key in instance.slots_keys(self):
                    if getattr(self, key) != getattr(other, key): return False
            
            return True
        except KeyError or AttributeError:
            return False
        
    def __neq__(self, other):
        return not self.__eq__(other)
    
    def __str__(self): return self.__repr__()
    
    def __repr__(self):
        return instance.slots_repr(self)
    
    def to_json(self):
        return {k:self._to_json(v) for k, v in instance.slots_items(self)}
    
    @classmethod
    def _to_json(cls, value):
        
        if type(value) in JSON_TYPES:
            return value
        elif hasattr(value, "to_json"):
            return value.to_json()
        elif isinstance(value, list):
            return [cls._to_json(v) for v in value]
        elif isinstance(value, dict):
            return {str(k):cls._to_json(v) for k,v in value.items()}
        else:
            raise TypeError("{0} is not json serializable.".format(type(value)))
    
    def __getstate__(self): return self.to_json()
    def __getnewargs__(self):
        return (self.to_json(), )
    
    #def __setstate__(self, doc): return self.intialize(**doc)

    @classmethod
    def from_json(cls, doc):
        return cls(**doc)
    
    _from_json = from_json


class AbstractJSONable(JSONable):
    """
    Implements a simple JSONable datastructure for abstract classes.
    
    :Example:
        >>> from jsonable import JSONable, AbstractJSONable
        >>>
        >>> class Bowl(JSONable):
        ...     __slots__ = ('fruit',)
        ...     def initialize(self, fruit):
        ...         self.fruit = [Fruit(f) for f in fruit]
        ...
        >>>
        >>>
        >>> class Fruit(AbstractJSONable):
        ...     __slots__ = ('weight',)
        ...     def initialize(self, weight):
        ...         self.weight = float(weight) # lbs
        ...
        >>>
        >>> class Apple(Fruit):
        ...     __slots__ = ('variety',)
        ...     def initialize(self, weight, variety):
        ...         super().initialize(weight)
        ...         self.variety = str(variety)
        ...
        >>> Fruit.register(Apple)
        >>>
        >>> class Orange(Fruit):
        ...     __slots__ = ('radius',)
        ...     def initialize(self, weight, radius):
        ...         super().initialize(weight)
        ...         self.radius = float(radius) # in
        ...
        >>> Fruit.register(Orange)
        >>>
        >>> orange = Orange(10.1, 2.5)
        >>>
        >>> apple = Apple(9.2, "Honey Crisp")
        >>>
        >>> bowl = Bowl([apple, orange])
        >>>
        >>> doc = bowl.to_json()
        >>>
        >>> bowl == Bowl(doc)
        True
    """
    __slots__ = tuple()
    
    REGISTERED_SUB_CLASSES = {}
    CLASS_NAME_KEY = "__class__"
    
    def to_json(self):
        doc = super().to_json()
        doc[self.CLASS_NAME_KEY] = self.__class__.__name__
        return doc
    
    @classmethod
    def from_json(cls, doc):
        if cls.CLASS_NAME_KEY in doc:
            class_name = doc[cls.CLASS_NAME_KEY]
            if class_name in cls.REGISTERED_SUB_CLASSES:
                SubClass = cls.REGISTERED_SUB_CLASSES[class_name]
            elif class_name == cls.__name__:
                SubClass = cls
            else:
                raise KeyError(str(class_name) +
                               " is not a recognized subclass of " +
                               cls.__name__)
            
            new_doc = copy.copy(doc)
            del new_doc[cls.CLASS_NAME_KEY]
            return SubClass.from_json(new_doc)
        else:
            return cls._from_json(doc)
    
    @classmethod
    def get(cls, class_name):
        return cls.REGISTERED_SUB_CLASSES[class_name]
    
    @classmethod
    def register(cls, SubClass):
        cls.REGISTERED_SUB_CLASSES[SubClass.__name__] = SubClass
